### PR TITLE
[Feature] Support Array Type compare function for nereids planner

### PR DESCRIPTION
**Pull request by koarz**
## Proposed changes

Issue Number: #31157

support Array Type Compare ( ">", "=","<",...)
```
mysql> select [1,2] = [1,2];
+-------------------+
| ([1, 2] = [1, 2]) |
+-------------------+
|                 1 |
+-------------------+
mysql> select [1,2] != [1,2];
+--------------------------+
| ( not ([1, 2] = [1, 2])) |
+--------------------------+
|                        0 |
+--------------------------+
mysql> select [1,3] > [1,2];
+-------------------+
| ([1, 3] > [1, 2]) |
+-------------------+
|                 1 |
+-------------------+
mysql> select [1,3] < [1,2];
+-------------------+
| ([1, 3] < [1, 2]) |
+-------------------+
|                 0 |
+-------------------+
mysql> select array_test.id, array_test.c_array, array_test.c_array = array_test1.c_array ,array_test1.id ,array_test1.c_array from array_test inner join array_test1;
+------+-----------------+---------------------+------+-----------------+
| id   | c_array         | (c_array = c_array) | id   | c_array         |
+------+-----------------+---------------------+------+-----------------+
|    1 | [1, 2, 3, 4, 5] |                   1 |    1 | [1, 2, 3, 4, 5] |
|    2 | [6, 7, 8]       |                   0 |    1 | [1, 2, 3, 4, 5] |
|    3 | []              |                   0 |    1 | [1, 2, 3, 4, 5] |
|    4 | NULL            |                NULL |    1 | [1, 2, 3, 4, 5] |
|    5 | [6, 7, 8]       |                   0 |    1 | [1, 2, 3, 4, 5] |
|    1 | [1, 2, 3, 4, 5] |                   1 |    1 | [1, 2, 3, 4, 5] |
|    2 | [4, 5]          |                   0 |    1 | [1, 2, 3, 4, 5] |
+------+-----------------+---------------------+------+-----------------+
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by koarz** [comment]
run buildall

File: fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java [hunk 1]
```diff
@@ -814,6 +814,9 @@ public TColumnType toColumnTypeThrift() {
      * here is to avoid special-casing logic in callers for concrete types.
      */
     public boolean matchesType(Type t) {
+        if (this.isArrayType() && t.isArrayType()) {
```
Comment (zclllyybb) [inline fe/fe-common/src/main/java/org/apache/doris/catalog/Type.java:817]: override in ArrayTypep

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run performance

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run performance

File: be/src/vec/columns/column_array.cpp [hunk 2]
```diff
@@ -89,7 +89,7 @@ INSTANTIATE_INDEX_IMPL(ColumnArray)
 
 ColumnArray::ColumnArray(MutableColumnPtr&& nested_column, MutableColumnPtr&& offsets_column)
         : data(std::move(nested_column)), offsets(std::move(offsets_column)) {
-    const ColumnOffsets* offsets_concrete = typeid_cast<const ColumnOffsets*>(offsets.get());
+    const auto* offsets_concrete = typeid_cast<const ColumnOffsets*>(offsets.get());
```
Comment (amorynan) [inline be/src/vec/columns/column_array.cpp:92]: why change this file ?

File: regression-test/suites/nereids_syntax_p0/array_compare.groovy [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+            SELECT * FROM ${tableName} where c_array = [];
+        """
+        order_qt_array_compare_3 """
+            SELECT * FROM ${tableName} where c_array != [1, 2, 3, 4];
+        """
+        order_qt_array_compare_4 """
+            SELECT * FROM ${tableName} where c_array > [1, 2, 3];
+        """
+        order_qt_array_compare_5 """
+            SELECT * FROM ${tableName} where c_array < [1, 2, 3, 4];
```
Comment (HappenLee) [inline regression-test/suites/nereids_syntax_p0/array_compare.groovy:55]: Add different type array compare

**Comment by amorynan** [comment]
why here no need to change file FunctionComparison.h to make be support work out ?

**Comment by koarz** [comment]
> why here no need to change file FunctionComparison.h to make be support work out ?

Because compare was implemented here before

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run buildall

**Comment by koarz** [comment]
run feut

**Review by HappenLee — DISMISSED** [review]
LGTM

File: regression-test/suites/nereids_syntax_p0/array_compare.groovy [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            SELECT * FROM ${tableName} where c_array > [1.0, 2.0, 3.0];
+        """
+        order_qt_array_compare_9 """
+            SELECT * FROM ${tableName} where c_array < [1.00001, 2.00001, 3.00001, 4.00001];
+        """
+    } finally {
+        try_sql("DROP TABLE IF EXISTS ${tableName}")
+    }
+
+    qt_sql """select [1, 2] = ["1, 2"];"""
```
Comment (amorynan) [inline regression-test/suites/nereids_syntax_p0/array_compare.groovy:73]: is here sql old planner support ? or just make 
```
sql """ set enable_nereids_planner = true;"""
        sql """ set enable_fallback_to_original_planner=false;"""
```  in code begin?

**Review by morrySnow — DISMISSED** [review]

File: regression-test/suites/nereids_syntax_p0/array_compare.groovy (hunk 4 above)
Comment (koarz) [reply regression-test/suites/nereids_syntax_p0/array_compare.groovy:73]: the sql not support on old planner

**Review by zclllyybb — APPROVED** [review]

**Comment by koarz** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by zclllyybb — APPROVED** [review]
LGTM

**Pull request merged**